2018/004-image-update-caasp-v4.md

```diff
@@ -0,0 +1,660 @@
+    version of the image exists.
+  * The new image is pulled from the registry because the intermediate tag
+    didn't change.
+  * The pod is started as soon as the image pull is completed.
+
+No downtime is going to be faced by the user even if the container engine has
+to pull the image from scratch. That happens because all our deployments are made
+by more than one replica, hence the requests are automatically routed to the
+other pods erogating the service.
+
+When the update cluster orchestration is done:
+
+  * The deployment definition will **not** change: the intermediate tag didn't
+    change.
+  * The nodes of the clusters are going to be rebooted due to the update of the
+    `<image>-metadata` package.
+  * The reboot of the nodes will lead kubernetes to schedule the pod somewhere
+    else on the cluster.
+  * The "always pull policy" will cause the new pod to start using the new image.
+
+The nodes are rebooted in a controlled fashion by our salt orchestration. Nodes
+are drained, hence no downtime is going to be experienced by the user.
+
+### Upgrade scenarios: static pods and daemon sets
+
+This section focuses on the update of the static pods (eg: HAproxy) and
+kubernetes daemon sets (eg: flannel and cilium).
+
+### New image is released
+
+This happens with one of the following cases:
+
+  * Core application version changes: new set of tags are created (short,
+    intermediate, long).
+  * Core application version doesn't change, a fix is done to the image
+    (eg: its entrypoint is fixed): short tag doesn't change, intermediate and
+    long tags change.
+
+Behind the scenes:
+
+  * [developers] Update KIWI definition of the image
+  * [developers] Update the `kubernetes-salt` package: the manifest file that
+    uses the image has to reference the new intermediate tag (which is constant
+    and always predictable).
+  * [QA] Test the fresh image. The stability and predictability of the
+    intermediate tags makes their life easier and reduces the chances of
+    mistakes.
+  * Image is published to the SUSE registry.
+  * `kubernetes-salt` and `<image>-metadata` packages are published to the
+    CaaSP Updates channel.
+
+Considerations about the images:
+
+  * The old image can still be pulled via its intermediate and long tags.
+  * Using the short tag will pull the new image.
+  * The new image can also be pulled using its intermediate and long tags.
+
+What happens when `transactional-update` is run:
+
+  * `transactional-update` notices there are package update
+    available on all the nodes of the cluster.
+  * `transactional-update` creates a new snapshot with the updated packages
+    installed and sets the "has updates" grain on each node of the cluster.
+  * Velum notices the value of the grain on the admin node: the
+    *"update admin node"* button is shown.
+  * Once the admin node is rebooted, Velum allows the user to press the
+    "update nodes" button
+
+**TODO:** answer these questions:
```

Comment:
@ereslibre: you have a better understanding of the upgrade orchestration. Can you help here?